src/Transform/ONNX/ConstProp.cpp

```diff
@@ -853,6 +860,150 @@ ONNXConstantOp ConstPropSlice(
+        std::vector<int64_t> inputIndices = getAccessIndex(k, inputStrides);
+        std::vector<int64_t> outputIndices(inputIndices);
+        outputIndices[axis] += dimAtAxis;
+        int64_t outputOffset =
+            getLinearAccessIndex(outputIndices, outputStrides);
+        if (elementType.isa<FloatType>()) {
+          double *inputArr = (double *)inputArrays[i];
+          double *outputArr = (double *)resArray;
+          *(outputArr + outputOffset) = *(inputArr + k);
+        } else if (elementType.isa<IntegerType>()) {
+          int64_t *inputArr = (int64_t *)inputArrays[i];
+          int64_t *outputArr = (int64_t *)resArray;
+          *(outputArr + outputOffset) = *(inputArr + k);
+        } else
+          llvm_unreachable("Unknown data type");
+      }
+      dimAtAxis += inputShape[axis];
+    }
+  }
+
+  // Construct a new ONNXConstantOp.
+  ONNXConstantOp res =
+      createConstantOpAndStoreBufferPtr(rewriter, replacingValue, resArray);
+
+  return res;
+}
+
+//===----------------------------------------------------------------------===//
+// Code to perform constant propagation for ExpandOp.
+//===----------------------------------------------------------------------===//
+
+ONNXConstantOp ConstPropExpand(
+    PatternRewriter &rewriter, Value replacingValue, Value constValue) {
+  // Get the const value using the maximum precision e.g. double, int64_t.
+  char *inputArray =
+      getArrayFromAttributeOrBuffer(rewriter, constValue.getDefiningOp());
+  // Create the result buffer using the maximum precision e.g. double, int64_t.
+  char *resArray =
+      allocateBufferFor(replacingValue.getType(), /*useMaxSize=*/true);
+
+  ArrayRef<int64_t> inputShape = getShape(constValue.getType());
+  std::vector<int64_t> inputStrides = getStrides(inputShape);
+  ArrayRef<int64_t> outputShape = getShape(replacingValue.getType());
+  std::vector<int64_t> outputStrides = getStrides(outputShape);
+  int64_t inputRank = inputShape.size();
+  int64_t outputRank = outputShape.size();
+  Type elementType = getElementType(replacingValue.getType());
+
+  // Check broadcasting.
+  bool broadcasting = false;
+  if (inputRank != outputRank)
+    broadcasting = true;
+  else
+    for (unsigned i = 0; i < outputRank; ++i)
+      if (inputShape[i] != outputShape[i]) {
+        broadcasting = true;
+        break;
+      }
+
+  for (int64_t i = 0; i < ShapedType::getNumElements(outputShape); ++i) {
+    // Compute indices to access the output.
+    std::vector<int64_t> outputIndices = getAccessIndex(i, outputStrides);
+    // Compute indices to access the input.
+    SmallVector<int64_t, 4> inputIndices(inputRank, 0);
+    if (!broadcasting) {
+      for (int k = 0; k < outputRank; ++k) {
+        inputIndices[k] = outputIndices[k];
+      }
+    } else {
+      for (int k = 0; k < outputRank; ++k) {
+        // in the input index range.
+        if (k >= outputRank - inputRank) {
```

Comment:
The broadcast rule counts dimensions from right to left, so `outputRank - inputRank` is the left offset for the input when inputRank < outputRank. For example,

```
output dimensions:   DimO1, DimO2, DimO3, DimO4, DimO5
input dimensions:                  DimI1, DimI2, DimI3
```

For each k-th dimension in the output, the corresponding input dimension will be `k - (outputRank - inputRank)`, so the condition `k >= (outputRank - inputRank)`

Looping over the output rank works for multiple inputs (This was has been using for binary ops). In this case, we have only one input so I can rewrite this loop using the input rank instead so that it is easy to understand.

Reply:
I rewrote the loop so that it is easy to understand. Thanks!